services/transactionService.ts

```diff
@@ -0,0 +1,249 @@
+      sort: 'desc',
+    },
+  });
+  const userRecentTransactions = result.data.result
+    .filter(tx => {
+      return tx.from.toLowerCase() === input.address.toLowerCase();
+    })
+    .map(tx => {
+      // in this case we know it's a token transfer (smart contract call)
+      return {
+        hash: tx.hash,
+        nonce: Number(tx.nonce),
+      };
+    });
+  return {
+    userRecentTransactions,
+    isTransactionListEmpty: result.data.result.length === 0,
+  };
+}
+
+async function getTransactionDetailForNormalTransfer(
+  input: TransactionDetailInput,
+): Promise<NetworkTransactionInfo | null> {
+  const { txHash, symbol, networkId } = input;
+  const transaction = await getNetworkWeb3(networkId).eth.getTransaction(
+    txHash,
+  );
+  if (!transaction) {
+    return null;
+  }
+  const block = await getNetworkWeb3(networkId).eth.getBlock(
+    transaction.blockNumber as number,
+  );
+
+  return {
+    from: transaction.from,
+    timestamp: block.timestamp as number,
+    to: transaction.to as string,
+    hash: txHash,
+    amount: normalizeAmount(transaction.value, 18),
+    currency: symbol,
+  };
+}
+
+async function getTransactionDetailForTokenTransfer(
+  input: TransactionDetailInput,
+): Promise<NetworkTransactionInfo | null> {
+  const { txHash, symbol, networkId } = input;
+  const token = findTokenByNetworkAndSymbol(networkId, symbol);
+  const web3 = getNetworkWeb3(networkId);
+  const transaction = await web3.eth.getTransaction(txHash);
+  console.log('getTransactionDetailForTokenTransfer', { transaction, input });
+  if (
+    transaction &&
+    transaction.to?.toLowerCase() !== token.address.toLowerCase()
+  ) {
+    throw new Error(
+      errorMessages.TRANSACTION_SMART_CONTRACT_CONFLICTS_WITH_CURRENCY,
+    );
+  }
+  if (!transaction) {
+    return null;
+  }
+
+  const transactionData = abiDecoder.decodeMethod(transaction.input);
+  const transactionToAddress = transactionData.params.find(
+    item => item.name === '_to',
+  ).value;
+  const block = await getNetworkWeb3(networkId).eth.getBlock(
+    transaction.blockNumber as number,
+  );
+  return {
+    from: transaction.from,
+    timestamp: block.timestamp as number,
+    hash: txHash,
+    to: transactionToAddress,
+    amount: normalizeAmount(
+      transactionData.params.find(item => item.name === '_value').value,
+      token.decimals,
+    ),
+    currency: symbol,
+  };
+}
+
+function validateTransactionWithInputData(
+  transaction: NetworkTransactionInfo,
+  input: TransactionDetailInput,
+): never | void {
+  if (transaction.to.toLowerCase() !== input.toAddress.toLowerCase()) {
+    throw new Error(
+      errorMessages.TRANSACTION_TO_ADDRESS_IS_DIFFERENT_FROM_SENT_TO_ADDRESS,
+    );
+  }
+
+  if (transaction.from.toLowerCase() !== input.fromAddress.toLowerCase()) {
+    throw new Error(
+      errorMessages.TRANSACTION_FROM_ADDRESS_IS_DIFFERENT_FROM_SENT_FROM_ADDRESS,
+    );
+  }
+  if (transaction.amount !== input.amount) {
```

Comment:
It's error-prone but cannot fix it by a simple solution. The `amount` data type is defined as `number` and changing it is laborous.